src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java

```diff
@@ -0,0 +1,219 @@
+            this.component = component;
+            this.length = length;
+        }
+
+        public String keyspace() throws KeyException
+        {
+            return Optional.ofNullable(keyspace).orElseThrow(KeyException::new);
+        }
+
+        public String table() throws KeyException
+        {
+            return Optional.ofNullable(table).orElseThrow(KeyException::new);
+        }
+
+        public String snapshot() throws KeyException
+        {
+            return Optional.ofNullable(snapshot).orElseThrow(KeyException::new);
+        }
+
+        public String component() throws KeyException
+        {
+            return Optional.ofNullable(component).orElseThrow(KeyException::new);
+        }
+
+        public KeyLength length()
+        {
+            return length;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o)
+            {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass())
+            {
+                return false;
+            }
+            Key key = (Key) o;
+            return length == key.length &&
+                   Objects.equals(keyspace, key.keyspace) &&
+                   Objects.equals(table, key.table) &&
+                   Objects.equals(snapshot, key.snapshot) &&
+                   Objects.equals(component, key.component);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(keyspace, table, snapshot, component, length);
+        }
+
+        /**
+         * Builder class for Key
+         */
+        public static class Builder
+        {
+            private String keyspace;
+            private String table;
+            private String snapshot;
+            private String component;
+            private int length;
+
+            public Builder setKeyspace(String keyspace)
+            {
+                length++;
+                this.keyspace = keyspace;
+                return this;
+            }
+
+            public Builder setTable(String table)
+            {
+                length++;
+                this.table = table;
+                return this;
+            }
+
+            public Builder setSnapshot(String snapshot)
+            {
+                length++;
+                this.snapshot = snapshot;
+                return this;
+            }
+
+            public Builder setComponent(String component)
+            {
+                length++;
+                this.component = component;
+                return this;
+            }
+
+            public CachedFilePathBuilder.Key build()
+            {
+                return new CachedFilePathBuilder.Key(keyspace, table, snapshot, component,
+                        KeyLength.values()[length - 1]);
+            }
+        }
+    }
+
+    /**
+     * Enum to hold key length values
+     */
+    public enum KeyLength
+    {
+        ONE, TWO, THREE, FOUR
```

Comment:
nit: it is only used to represent 1, 2, 3, 4... declaring enum types seem overly complicated.

Reply:
Mainly added to improve readability of this switch statement https://github.com/apache/cassandra-sidecar/blob/0083e0d62dd0a9aa63fa977ffeea6d7609f387c3/src/main/java/org/apache/cassandra/sidecar/utils/CachedFilePathBuilder.java#L76

Reply:
Numbers are not best for readability. Please declare the enum types that are more descriptive. 
For example, `ONE` represents `JUST_KEYSPACE`, `TWO` means `KEYSPACE_TABLE`, `THREE` is `KEYSPACE_TABLE_SNAPSHOT`, so on so forth.